tests/learning/test_fault_tolerance.py

```diff
@@ -0,0 +1,191 @@
+from functools import partial
+
+import pytest
+from apistar.http import Response
+from eth_keys.datatypes import Signature as EthSignature
+from twisted.logger import globalLogPublisher, LogLevel
+
+from constant_sorrow.constants import NOT_SIGNED
+from nucypher.characters.lawful import Ursula
+from nucypher.characters.unlawful import Vladimir
+from nucypher.crypto.powers import SigningPower, CryptoPower
+from nucypher.network.nodes import FleetStateTracker
+from nucypher.utilities.sandbox.constants import INSECURE_DEVELOPMENT_PASSWORD
+from nucypher.utilities.sandbox.middleware import MockRestMiddleware
+from nucypher.utilities.sandbox.ursula import make_federated_ursulas
+
+
+def test_blockchain_ursula_substantiates_stamp(blockchain_ursulas):
+    first_ursula = list(blockchain_ursulas)[0]
+    signature_as_bytes = first_ursula._evidence_of_decentralized_identity
+    signature = EthSignature(signature_bytes=signature_as_bytes)
+    proper_public_key_for_first_ursula = signature.recover_public_key_from_msg(bytes(first_ursula.stamp))
+    proper_address_for_first_ursula = proper_public_key_for_first_ursula.to_checksum_address()
+    assert proper_address_for_first_ursula == first_ursula.checksum_public_address
+
+    # This method is a shortcut for the above.
+    assert first_ursula._stamp_has_valid_wallet_signature
+
+
+def test_blockchain_ursula_verifies_stamp(blockchain_ursulas):
+    first_ursula = list(blockchain_ursulas)[0]
+
+    # This Ursula does not yet have a verified stamp
+    first_ursula.verified_stamp = False
+    first_ursula.stamp_is_valid()
+
+    # ...but now it's verified.
+    assert first_ursula.verified_stamp
+
+
+def test_blockchain_ursula_is_not_valid_with_unsigned_identity_evidence(blockchain_ursulas, caplog):
+    lonely_blockchain_learner, blockchain_teacher, unsigned = list(blockchain_ursulas)[0:3]
+
+    unsigned._evidence_of_decentralized_identity = NOT_SIGNED
+
+    # Wipe known nodes.
+    lonely_blockchain_learner._Learner__known_nodes = FleetStateTracker()
+    lonely_blockchain_learner._current_teacher_node = blockchain_teacher
+
+    lonely_blockchain_learner.remember_node(blockchain_teacher)
+    warnings = []
+
+    def warning_trapper(event):
+        if event['log_level'] == LogLevel.warn:
+            warnings.append(event)
+
+    globalLogPublisher.addObserver(warning_trapper)
+
+    lonely_blockchain_learner.learn_from_teacher_node()
+
+    globalLogPublisher.removeObserver(warning_trapper)
+
+    # We received one warning during learning, and it was about this very matter.
+    assert len(warnings) == 1
+    assert warnings[0]['log_format'] == unsigned.invalid_metadata_message.format(unsigned)
+
+    assert len(lonely_blockchain_learner.known_nodes) == len(
```

Comment:
I know we dont have a length convention established, but why break on `(`?

Reply:
Yeeeahhh, that'd be the old CTRL-ALT-L.  I'll fix it.